File: 추천50제/bfs/2146.py
import queue
import sys
from collections import deque
n = int(sys.stdin.readline())
board = []
for _ in range(n):
    board.append(list(map(int,sys.stdin.readline().split())))

#각 섬마다 좌표 모두 저장후
#완탐으로 좌표별 거리 빼서 최단 값 구하기

dx = [0,0,-1,1]
dy = [1,-1,0,0]
stack = []
answer = []
for i in range(n):
    for j in range(n):
        if board[i][j]==1:
            temp = [[i,j]]
            queue = deque([[i,j]])
            while queue:
                x,y = queue.popleft()
                for k in range(4):
                    nx = x+dx[k]
                    ny = y+dy[k]
                    if -1<nx<n and -1<ny<n and board[nx][ny]==1:
                        queue.append([nx,ny])
                        board[nx][ny]=0
                        temp.append([nx,ny])
            stack.append(temp)

for i in range(len(stack)):
    for j in range(i+1,len(stack)):
        for k in stack[i]:
            for z in stack[j]:
                answer.append(abs(k[0]-z[0])+abs(k[1]-z[1]))
print(min(answer)-1)

File: BOJ/2146.js
let fs = require('fs');
let input = fs
  .readFileSync(process.platform == 'linux' ? './dev/stdin' : './test.txt')
  .toString()
  .trim()
  .split('\n');
let n = parseInt(input.shift());
let board = input.map((e) => e.split(' ').map(Number));

let dx = [-1, 1, 0, 0];
let dy = [0, 0, -1, 1];

let answer = [];
let islands = [];
let i_number = 1;
const bfs = (x, y, i_number) => {
  let temp = [[x, y, 0]];
  let q = [[x, y]];
  board[x][y] = i_number;
  while (q.length) {
    let [x, y] = q.shift();
    for (let i = 0; i < 4; i++) {
      let nx = x + dx[i];
      let ny = y + dy[i];
      if (-1 < nx && nx < n && -1 < ny && ny < n && board[nx][ny] == 1) {
        board[nx][ny] = i_number;
        q.push([nx, ny]);
        temp.push([nx, ny, 0]);
      }
    }
  }
  return temp;
};

for (let i = 0; i < n; i++) {
  for (let j = 0; j < n; j++) {
    if (board[i][j] == 1) {
      islands.push(bfs(i, j, ++i_number));
    }
  }
}

const search = (q, number) => {
  let temp = [];
  let visited = Array.from({ length: n }, () => new Array(n).fill(0));
  visited[q[0][0]][q[0][1]] = 1;
  while (q.length) {
    let [x, y, distance] = q.shift();
    for (let i = 0; i < 4; i++) {
      let nx = x + dx[i];
      let ny = y + dy[i];
      if (-1 < nx && nx < n && -1 < ny && ny < n && board[nx][ny] != number && visited[nx][ny] == 0) {
        visited[nx][ny] = 1;
        if (board[nx][ny] != 0 && board[nx][ny] != number) {
          //다른 육지이면
          temp.push(distance + 1);
        } else {
          q.push([nx, ny, distance + 1]);
        }
      }
    }
  }
  answer.push(Math.min(...temp));
};
islands.forEach((e, index) => {
  search(e, index + 2);
});
console.log(Math.min(...answer) - 1);
